Client side `on-demand` Sync Mode (#6)
powersync-ja/temp-tanstack-db: packages/powersync-db-collection/sync-stream-examples.md

> ## Calling sync streams automatically

Reviewer: ```suggestion
## Calling Sync Streams automatically
```

> For the these examples we assuming the follow sync stream exists:

Reviewer: ```suggestion
For the these examples we are assuming the follow Sync Streams exists:
```

```
streams:
  lists:
    query: SELECT * FROM lists WHERE owner_id = auth.user_id()
    auto_subscribe: true
  todos:
    query: SELECT * FROM todos WHERE list_id = subscription.parameter('list') AND list_id IN (SELECT id FROM lists WHERE owner_id = auth.user_id())

config:
  edition: 2
```

### Example 1: Eager mode basic usage

```typescript
const collection = createCollection(
  powerSyncCollectionOptions({
    database: db,
    table: AppSchema.props.todos,

>     syncMode: `eager`,

Reviewer: I noticed sometimes we use backticks for syncMode, sometime single quotes

## Incorporating Sync Streams

Ideally we would be able to map TanstackDB queries to sync streams automatically, if we can optimise the amount of data sync to
the sqlite database from the service we have smaller set of data that needs to be considered when syncing from the sqlite database to TanstackDB collections.

As a stepping stone towards that, we now expose data loading hooks for both eager and on-demand sync modes that allow a user to call sync streams when a collection is defined (eager mode) or when a collection's data boundary changes based on the live queries predicates (on-demand).
For the these examples we are assuming the follow sync stream exists:

```
config:
  edition: 3

streams:
  lists:
    query: SELECT * FROM lists WHERE owner_id = auth.user_id()
    auto_subscribe: true
  todos:
    query: SELECT * FROM todos WHERE list_id = subscription.parameter('list') AND list_id IN (SELECT id FROM lists WHERE owner_id = auth.user_id())
```

### Example 1: Eager mode basic usage

> If you want an eager collection to subscribe to a sync stream when a collection loads, you can use the `onLoad` hook.

Reviewer: ```suggestion
Use the `onLoad` hook to subscribe to a Sync Stream when the collection first loads, so SQLite is populated before the collection starts serving queries. In `eager` mode, all rows in SQLite are synced into the TanStack DB collection; live query filtering then runs against that full dataset. 
```

> The hook may return a cleanup function.

Reviewer: ```suggestion
The hook can optionally return a cleanup function where you can unsubscribe from the Sync Stream.
```

Consider the diagram as an example.
We start with 4 todos in the PS service, only 2 todos get synced via the sync stream to the SQLite database. Because it's eager mode, both get synced from the SQLite database to the collection. Finally the TanstackDB query only returns the single todo that matches the live query predicate.

```typescript
const collection = createCollection(
  powerSyncCollectionOptions({
    database: db,
    table: AppSchema.props.todos,
    syncMode: 'eager',
    onLoad: async () => {
      console.log('onLoad')
      const subscription = await db
        .syncStream('todos', { list: 'list_1' })

>         .subscribe({ ttl: 0 })

Reviewer: ```suggestion
        .subscribe({ ttl: 0 }) // TTL 0 means data is evicted immediately on unsubscribe
```

Reviewer: I assume the TTL: 0 here is deliberate? So a brief explainer in these examples would be useful

Author: It's a preference thing, I also like to use it for testing. I'll remove it.

      await subscription.waitForFirstSync()

      return () => {
        console.log('onUnload')
        subscription.unsubscribe()
      }
    },
  }),
)
```

> A live query that filters by completed.

Reviewer: ```suggestion
A live query that filters by the `completed` state.
```

```typescript
const liveQuery = createLiveQueryCollection({
  query: (q) =>
    q
      .from({ todo: collection })
      .where(({ todo }) => eq(todo.completed, 1))
      .select(({ todo }) => ({
        id: todo.id,
        completed: todo.completed,
      })),
})
```

### Example 2: On-demand basic usage

> If you want to on-demand collection to subscribe to a sync stream whenever a subset of data is loaded (when the list of live queries against the collection change), you can use the `onLoadSubset` hook.

Reviewer: ```suggestion
Use the `onLoadSubset` hook to subscribe to a Sync Stream whenever the collection's data boundary changes (i.e. when the set of active live queries changes). In `on-demand` mode, only rows that satisfy the active live query predicates are synced from SQLite into the TanStack DB collection. 
```

> The hook may return a cleanup function.

Reviewer: ```suggestion
The hook can optionally return a cleanup function where you can unsubscribe from the Sync Stream when that subset is no longer needed.
```

Consider the diagram as an example.

> We start with 4 todos in the PS service, only 2 todos get synced via the sync stream to the SQLite database. Because it's on-demand mode, only 1 todo matches gets synced from the SQLite database to the collection. Finally the TanstackDB query only returns the single todo that matches the live query predicate.

Reviewer: PS service should be "PowerSync Service"
